rider-fsharp/src/main/java/com/jetbrains/rider/plugins/fsharp/services/fsi/FsiHost.kt

```diff
@@ -1,22 +1,119 @@
+import com.intellij.execution.process.ProcessAdapter
+import com.intellij.execution.process.ProcessEvent
 import com.intellij.notification.Notification
 import com.intellij.notification.NotificationType
 import com.intellij.notification.Notifications
 import com.intellij.openapi.editor.Editor
 import com.intellij.openapi.project.Project
-import com.intellij.openapi.util.io.FileUtil
+import com.intellij.openapi.project.guessProjectDir
+import com.intellij.openapi.util.NlsSafe
+import com.intellij.openapi.vfs.VfsUtil
+import com.intellij.project.isDirectoryBased
 import com.intellij.psi.PsiFile
-import com.jetbrains.rdclient.util.idea.LifetimedProjectComponent
-import com.jetbrains.rider.projectView.solution
+import com.intellij.util.ui.UIUtil
 import com.jetbrains.rd.util.reactive.Property
 import com.jetbrains.rd.util.reactive.flowInto
+import com.jetbrains.rdclient.util.idea.LifetimedProjectComponent
 import com.jetbrains.rider.plugins.fsharp.*
+import com.jetbrains.rider.plugins.fsharp.services.fsi.consoleRunners.FsiConsoleRunnerBase
+import com.jetbrains.rider.plugins.fsharp.services.fsi.consoleRunners.FsiDefaultConsoleRunner
+import com.jetbrains.rider.plugins.fsharp.services.fsi.consoleRunners.FsiScriptProfileConsoleRunner
+import com.jetbrains.rider.plugins.fsharp.services.fsi.runScript.FSharpScriptConfiguration
+import com.jetbrains.rider.projectView.solution
+import com.jetbrains.rider.run.configurations.dotNetExe.DotNetExeConfigurationParameters
+import com.jetbrains.rider.run.createRunCommandLine
+import com.jetbrains.rider.runtime.DotNetExecutable
+import com.jetbrains.rider.runtime.DotNetRuntime
+import com.jetbrains.rider.runtime.RiderDotNetActiveRuntimeHost
+import com.jetbrains.rider.runtime.dotNetCore.DotNetCoreRuntimeType
+import com.jetbrains.rider.runtime.mono.MonoRuntime
+import com.jetbrains.rider.runtime.mono.MonoRuntimeType
+import com.jetbrains.rider.runtime.msNet.MsNetRuntimeType
 import org.jetbrains.concurrency.AsyncPromise
 import org.jetbrains.concurrency.Promise
+import org.jetbrains.concurrency.resolvedPromise
+import java.awt.event.FocusEvent
+import java.awt.event.FocusListener
+import java.util.*
+
+fun getFsiRunOptions(
+  project: Project,
+  configuration: FSharpScriptConfiguration? = null
+): Pair<DotNetExecutable, DotNetRuntime> {
+  val fsiHost = project.getComponent(FsiHost::class.java)
+  lateinit var sessionInfo: RdFsiSessionInfo
+  UIUtil.invokeLaterIfNeeded {
+    sessionInfo = fsiHost.rdFsiHost.requestNewFsiSessionInfo.sync(Unit)
+  }
+
+  val runtimeHost = RiderDotNetActiveRuntimeHost.getInstance(project)
+
+  val exePath = when (sessionInfo.runtime) {
+    RdFsiRuntime.Core -> runtimeHost.dotNetCoreRuntime.value!!.cliExePath
+    RdFsiRuntime.NetFramework -> sessionInfo.fsiPath
+    RdFsiRuntime.Mono -> {
+      val runtime = runtimeHost.getCurrentClassicNetRuntime(false).runtime as MonoRuntime
+      runtime.getMonoExe().path
+    }
+  }
```

Comment:
I think we have a helper function for the runtime itself to append its executable if required; have you considered using that?

Reply:
Can you tell me what utility is? I noticed some that modified `GeneralCommandLine`, but we need a utility to create and validate DotnetExecutable exe.

Reply:
Aw yeah, so far it is only `DotNetRuntime::patchRunCommandLine(commandLine: GeneralCommandLine, runtimeArguments: List<String>)`, so it is not directly applicable here.

I would recommend refactoring it in Rider core to apply here as well. But so far your solution may live.